### ISSUE TITLE
External orchestrator to manage positions (multi-strategies in live)

### ISSUE THREAD

**Issue by gaugau3000** [freqtrade/freqtrade#3604]
Hi,

When you have two strategies or more in live they are a kind of isolated and cannot talk to each other (witch is normal).
But to manage the risk you may would like to allow a max open trade, not buy the same pair... at the multi-strategy/system level (because of hyper correlation in crypto and especially during flash crash/pump) and not only in the strategy.

I am pretty interested on how you deal with this in freqtrade ?

Have a nice day.

**Comment by gaugau3000** [comment]
As a trader
I want to use an external API orchestrator to manage the positions
So that i can configure the risk at multi strategy level (max number trades open, different stake amount....)

Don't want to go too deep for now but smt like : 
```
{
max_open_trades : 'external',
stake_amount:'external',
orchestrator : {
   enable : true,
   ...
}
}
```

Implementation of confirm_trade_entry method in IStrategy :
Hello external api orchestrator my strategy name is 'GoToTheMoon' and my pair is 'ADA/USDT' can you please tell me if i can make my buy and if so the stake amount associated.

Have a nice day.

**Comment by xmatthias** [comment]
Hi,

Implementing an external API orchestrator may sound like a good idea at first, but it's not something we should have as part of freqtrade.

It will complicate both the code (increase the maintenance burden) - as well as complicate and raise the entry-barrier for new users.

The main reason i think it's a bad idea is security (and complexity).

In the simplest case, we're talking about 2 bots running on the same machine. 

However, if you run into API limit issues (-> you're using kraken) - you'll need to deploy the bots on multiple machines (most likely in the cloud), so each machine only runs one bot.

Now we need to think about how to secure the connection between 2 cloud machines, and how to let the bots know how to reach the api orchestrator ...

Sure, you can connect the 2 machines via a P2P VPN - or SSH tunnels ... but this is not something we have the ressources to support people with - and looking at our average userbase (based on active users on slack ...), i think >75% will not have the knowledge / skills to configure / research this themselfs (which will result in them asking for support anyway). (to be clear - it's not that these are not clever users - but users with no IT technical background).

Even assuming you can come up with a clever way to make the above easy for non-it people ....
Now i want a 3rd bot (on a 3rd machine) - this will complicate the setup again by a dimension (-> this will not really scale).

Even worse - i've now got my 5 bots configured - but my orchestrator goes down / has a problem - this will stop all 5 bots. Now this must either be handled within freqtrade (reconnect if it's gone) - or i need to go to all 5 machines to restart freqtrade (which i must do anyway, assuming the orchestrator machine had it's address changed for whatever reason).

For all of the above problems there are technical solutions of some sort  - but none of which a non-technical user will be able to setup without requiring major help (it'll be different if your'e using AWS, or DigitalOcean, or Hetzner, or some other provider)..


This is possible to setup within an organization (where you can build a firewall around all servers - but the servers can comunicate freely between each other - or even in the cloud, where you build Private network within the cloud) - but to provide this as feature to (non-techincal) users will cause more problems than it'll solve - it'll result in more servers on the internet beeing attackable - which is something i don't want to be responsible for by even documenting this as a possibility.




To summarize, i see the following problems:
* secure the connection between different machines (in a way inexperienced users will be able to configure it without the risk of making a mistake)
* ensure fault tolerance (if the orchestrator is gone / changes)
* low adoption and usage since the setup and configuration will be difficult (this will result in the freqtrade team wasting a lot of time on a edgecase solution for 2-3 people)

**Comment by gaugau3000** [comment]
Hi,

I understand all of your points. You seems to focus straight on the "how" and less on the "why" :-) . 

In term of risk management  run 2 strategies in live with freqtrade at the same time is not a good idea. If you are in multiple trades at the same time and BTC lose 50% of his value (it seems it happens sometimes ;-) ) then your drawdown/profit ratio will be terrible (in my point of view this is the best metric to see risk/reward and doing sizing associated). 

Of course all your technical points are valid.

I know not everyone need this kind of tool.

At some point i will need to build an orchestration tool and i think maybe it is a good idea to open source it so my reflection is about how i can integrate it with freqtrade ?

Maybe for the 'client part' of this kind of project allow freqtrade to have extensions (i really don't know the how ? ) that mean this part is not maintained by freqtrade team but the users are able to use it and share ideas and i host the orchestration tool ?

Have a nice day.

PS : i am on the early stage of thinking so not everything is clear in my head.

**Comment by xmatthias** [comment]
If BTC dumps 50%, your drawdown will be terrible no matter how you try and protect - as almost all pairs are highly correlated to BTC.


Also, it's not about hosting (i guess most people would want to host it themselfs anyway...) ... but about maintaining it as part of the freqtrade codebase.

If it's an external part / tool, we can also have it as part of the freqtrade organization (so it's clear it's meant to be used together with freqtrade). 

The above questions (how to secure the stuff) must still be solved (that would also be the case if you host it). There are solutions - i'm just not sure many are suitable for the average user.

---

> Hello external api orchestrator my strategy name is 'GoToTheMoon' and my pair is 'ADA/USDT' can you please tell me if i can make my buy and if so the stake amount associated.

Technically, you can do this purely within the strategy using the new [`confirm_trade_entry` hook ](https://www.freqtrade.io/en/latest/strategy-advanced/#trade-entry-buy-order-confirmation).


Now to avoid having users do this manually (and make mistakes...), you could try the following  idea i had recently, while thinking about how to allow users to easily enable "standard" options, which however can be done via a callback.

i could imagine a Mixin (so your strategy class would look something like this):

```
from somewhere import ExtOrch
# ...

class Awesomestrategy(ExtOrch, IStrategy):

    def __init__(self, ...):
        self.init_Orchestrator(*whateverargslikepassword)

```

This would then require the strategy to NOT overwrite confirm_trade_entry (as this is provided by the ExtOrch mixin) and does some "standard" things.

Obviously, users could still overwrite this method and call implement additional checks  - but i guess that's details then.

**Comment by xmatthias** [comment]
I'll close this due to inactivity. 

Feel free to reopen if you want to discuss this further - but for now i don't think this will be integrated into freqtrade (which was also never really the plan in the first place).
